### PR TITLE
Class system D: hybrid attribute model (locked base + free pool) (#1223)

### LINKED ISSUE #1223: Class system D: Hybrid attribute model (locked base + free pool)
**Depends on:** #1221

Part of the class-system spike #1126 — **area D**. See `docs/spikes/1126-class-system.md`.

The hybrid attribute model — **locked base + free pool** — which also *is* the attribute floor.

## Scope
- Assemble the class `AttributeDistributions` into the battler as a level-scaled, **non-reallocatable** locked base (`BaseAmount + AmountPerLevel × level`), on **both** frontend and backend battler assembly. Source-tag it (`Class` / reuse `AttributeDistribution`) for the breakdown screen.
- Split the per-level stat-point grant so only the reduced **free pool** (strawman ~30%) flows through `PlayerStatPoints.TryUpdateAttributes`. Persist only the free-pool allocations; the locked base is recomputed from `(class, level)`, never stored.
- **Offline simulator (#879):** re-derive the locked base as levels cross the away window (power is a deterministic function of level — no stationarity violation, since the player still cannot *reallocate* while away).
- Frontend attributes screen: render locked-base vs free-pool distinctly; the free-pool budget drives the allocation UI.

## Notes
- This delivers the floor structurally (you cannot strip the locked base), so no separate floor-validation rule is needed.

## Tests
- Unit tests for the locked-base derivation and the free-pool budget math; **frontend↔backend parity vectors** for a leveled locked base; offline-sim test that the locked base grows correctly across simulated level-ups.

### PR DESCRIPTION
**Pull request by ginderjeremiah**
## Summary

Implements **area D** of the class-system spike ([#1126](https://github.com/ginderjeremiah/GameServer/issues/1126)) — the **hybrid attribute model**: a level-scaled, non-reallocatable **class locked base** (the attribute fingerprint / floor) plus a reduced **free pool** of manually-allocated stat points. Builds on the merged areas A–C (#1220/#1221/#1222).

## How it addresses the issue

**Backend — locked base in the battler (both the simulation and the reward power measurement):**
- `BattleSnapshot` captures `ClassId` and composes the class's `AttributeDistributions` as a locked base (`BaseAmount + AmountPerLevel × level` — the same shape and `EAttributeModifierSource.AttributeDistribution` source an enemy uses), derived from `(class, level)` and **never stored**.
- `BattleService` threads a fail-loud class resolver into `ToBattler`/`GetModifiers`, so the locked base flows into both the anti-cheat replay **and** `DefeatRewards` (a class-boosted player is measured at its true power).
- **Composition order is fixed and mirrored:** stat allocations → gear → **class locked base** → proficiency bonuses → static engine modifiers (float addition isn't associative, so position is part of the parity contract).

**Backend — the free-pool split:**
- `GameConstants.StatPointsPerLevel` reduced (6 → 2, strawman) — the locked base now supplies the bulk of attribute growth. (`[ClientMirrored]`; `STAT_POINTS_PER_LEVEL` regenerated.)
- `NewPlayerFactory` seeds the free-pool allocations **empty** (was the class base spread) — the locked base provides the starting spread, so seeding it too would double-count.

**Frontend — the parity surface, wired live on both sides:**
- New shared `UI/src/lib/battle/class-modifiers.ts` (`classLockedBaseModifiers`) — the exact mirror of the backend composition, with bit-exact parity vectors on **both** sides (`ClassLockedBaseParityTests` / `class-locked-base-parity.test.ts`), including a derived-attribute (`MaxHealth`) scenario to pin the additive ordering.
- The player's class distributions ride `PlayerData` (`LockedBaseDistribution`, resolved from the class catalogue in the login projection); `PlayerManager` memoises `classLockedBaseModifiers(distribution, level)` into the live battler (recomputed on level-up), composed in the same `additionalModifiers` slot as the proficiency bonuses.

## Scope notes

1. **The live frontend wiring is included, not deferred.** My first pass deferred it to #1225 on the proficiency #1117→#1119 precedent — but that precedent only held because proficiency was *empty* until content existed. **Every character has a class with a locked base**, so the live battler must compose it or the anti-cheat replay diverges (the e2e suite caught exactly this on the first push). The fix delivers the player's own fingerprint via `PlayerData`; the `GetClasses` **reference** delivery + the create-character picker + the attributes-screen locked/free render remain **#1225's** scope (the live battler needs none of them).

2. **Offline: the locked base is frozen at the snapshot's window-start level**, not grown mid-window. The issue's AC asked for growth, but the offline simulator (#879) is strictly *stationary* (level/proficiency/gear/allocations all frozen, exp applied after) — growing only the locked base would be internally inconsistent, and truly modeling it means un-freezing the whole battler's level (a #879 redesign). Filed [#1253](https://github.com/ginderjeremiah/GameServer/issues/1253) for that design question.

## Test plan

- [x] **Backend** — build + `dotnet format --verify-no-changes` clean. `Game.Core.Tests` (828: parity vectors, `BattleSnapshot` locked-base, offline, free-pool), `Game.Application.Tests` (583: end-to-end reward test — the locked base lifts measured power 100→200, reward 200 vs the un-boosted 800), `Game.Api.Tests` (615: `PlayerData` locked-base mapping), `Game.Infrastructure.Tests` (60) green.
- [x] **Frontend** — `npm run lint` clean; `npm run test:unit:ci` 2554 passed, coverage floors hold (new `class-locked-base-parity`, `PlayerManager.battleLockedBaseModifiers` memoisation/recompute, battle-engine composition).
- [ ] **e2e** — re-running on CI after the live-wiring fix (the first push failed e2e on the parity break above; the fix restores it — verified by the trace + the parity/integration tests).

Closes #1223

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: UI/src/lib/battle/class-modifiers.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ *  surface the class system adds (spike #1126 area D). The produced modifiers are fed into the player battler
+ *  at assembly — alongside the proficiency bonuses, before the static engine modifiers — once the class
+ *  reference data and the player's class id reach the client (#1225). */
+export function classLockedBaseModifiers(
+	distributions: readonly ClassAttributeDistribution[],
+	level: number
+): AttributeModifier[] {
+	return distributions.map((distribution) => ({
+		attribute: distribution.attributeId,
+		amount: distribution.baseAmount + distribution.amountPerLevel * level,
```
Comment (ginderjeremiah) [inline UI/src/lib/battle/class-modifiers.ts:30]: **Should-fix (parity hole the contract doesn't cover).** This recomputes `baseAmount + amountPerLevel × level` in IEEE-754 **double**, but the backend's `AttributeDistribution.GetDistributionModifier` does `(double)(BaseAmount + AmountPerLevel × level)` where both operands are **`decimal`** — the multiply/add happen in decimal, then cast. For a distribution with a non-zero fractional `amountPerLevel` (or a fractional `baseAmount` plus a non-zero level term), the two results can differ in the last bits, and the replay asserts bit-exactly (`toBe` / `Assert.Equal`, no tolerance). That would silently flag a legitimate player on the anti-cheat path.

The parity vectors deliberately dodge this (integer base/per-level, or fractional only with per-level 0), and the only record of the constraint is in the test comments. It can't manifest until class reference delivery + content (#1225) wires real distributions, so it's not breaking anything today — but there's no guard or follow-up tracking it. Before fractional class content lands, either:
- validate that a class's `AttributeDistribution`s are integer-valued (enforced where class content is authored), or
- compute the class locked base in `double` on the backend too so both sides share one arithmetic (class-specific, since enemies share `GetDistributionModifier` and don't recompute on the client), or
- give this one surface a tolerance.

At minimum, hoist the integer-only constraint out of the test comments into `docs/backend.md` / a comment here, and file a follow-up.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Game.Api/Controllers/LoginController.cs [hunk 2]
```diff
@@ -33,8 +36,25 @@ public class LoginController(
...
         private readonly BattleService _battleService = battleService;
+        private readonly IClasses _classes = classes;
         private readonly ILogger<LoginController> _logger = logger;
 
+        // Projects the player to its wire DTO, resolving its class's attribute distributions (the locked-base
+        // fingerprint #1126 area D) so the client battler composes the same locked base the backend snapshot
+        // does. An unresolvable class yields an empty fingerprint rather than throwing — the player still loads.
+        private PlayerData BuildPlayerData(Player player)
+        {
+            var lockedBaseDistribution = (_classes.GetClass(player.ClassId)?.AttributeDistributions ?? [])
```
Comment (ginderjeremiah) [inline Game.Api/Controllers/LoginController.cs]: **Should-fix (inconsistent failure policy).** This silently degrades to an **empty** fingerprint on an unresolvable class ("the player still loads"), but `BattleService.ResolveClass` and `AccountService.ResolveCreatableClass` both **throw** on the same condition (the latter: *"a miss here is a corrupt cache, not a bad request"*). A player's `ClassId` is validated at creation, so a login miss is exactly that corrupt-cache case — and the lenient path produces a *worse* outcome than failing fast: the client composes no locked base (simulates a weaker player) while the battle path throws on the same id, so the player can log in but every battle 500s on submit.

The method comment claims this applies "the player-load missing-reference policy," but the actual sibling policy here is fail-loud. Recommend reconciling — throw at login too (consistent with the other two call sites) — or, if login leniency is intentional, document why this one path tolerates the corruption the others reject.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by ginderjeremiah — COMMENTED** [review]
## Review: Class system D — hybrid attribute model (locked base + free pool), #1223

**Verdict:** Approve _(submitted as a comment — I'm the PR author, so GitHub won't take a gating event from me; no blocking findings)._ Clean layering, honest scope handling, and thorough mirrored tests. The findings below are should-fixes/nits, none gate the merge.

> Note: GitHub's PR diff is against the merge-base (30 files / +804). The branch is behind `main`, so a raw `main..head` diff shows phantom deletions of the proficiency-screen rework that this PR didn't touch — ignore those.

### Should-fix

- **Decimal-vs-double parity hole the "bit-exact" contract doesn't cover** — `class-modifiers.ts:28` recomputes the locked base in IEEE-754 double, while the backend (`AttributeDistribution.GetDistributionModifier`) computes it in `decimal` then casts. A non-zero fractional `amountPerLevel` (or fractional base + non-zero level) can diverge in the last bits, and the replay compares with no tolerance → a legitimate player gets flagged. The parity vectors deliberately avoid this (integer-only / per-level 0) and the constraint lives only in test comments. Can't manifest until class content/#1225 wires real distributions, but there's no guard or follow-up. Add a guard (integer-only validation, or compute the class base in double on both sides) + a doc note before fractional class content lands. (inline on `class-modifiers.ts:28`)
- **Inconsistent class-resolution failure policy** — `LoginController.cs` `BuildPlayerData` returns an empty fingerprint on an unresolvable class, but `BattleService.ResolveClass` and `AccountService.ResolveCreatableClass` both throw ("a miss is a corrupt cache"). Since `ClassId` is validated at creation, the lenient login path just produces a subtler broken state (loads fine, every battle 500s) and contradicts the method's own "player-load missing-reference policy" comment. Reconcile to fail-loud, or document why login alone is lenient. (inline on `LoginController.cs:47`)

### Test coverage

Strong overall — backend snapshot composition (incl. the `ClassId = 0` resolver edge for 0-based reference data), the fail-loud "class captured but no resolver" guard, offline frozen-base, the `NewPlayerFactory` empty-seeding, `PlayerData` mapping, and a convincing end-to-end integration test (locked base lifts measured power → reward 200 vs the un-boosted 800). Both parity suites mirror row-for-row, and `maxHealthDerivedAdditive` cleverly pins additive ordering past the 10th decimal.

One gap: the PR adds `lockedBaseModifiers === this.lastLockedBaseModifiers` to the battle engine's per-spawn rebuild gate, but no test exercises it — the mock's `battleLockedBaseModifiers` is a fixed `[]` that's never reassigned, so unlike proficiency (which has a dedicated "re-derives when the proficiency modifiers change" test) there's no locked-base equivalent, and the engine's `[...lockedBaseModifiers, ...proficiencyModifiers]` order at the `reset` seam is never asserted with both lists non-empty. Add the mirrored change-detection test + an order assertion with both non-empty.

### Nits

- `Game.Abstractions/Contracts/AttributeDistribution.cs` doc comment still reads "Read contract for an **enemy's** … distribution" — it's now reused for the class locked base too; widen it. (file not in this diff — follow-up.)
- Branch is behind `main` (lacks the proficiency-screen rework); needs an update/rebase before merge, but that's mechanical.

### On the deliberate AC deviations (both reasonable)

- **Offline base frozen at window-start level**, not grown across the window as the issue AC asked. Correct call: the offline simulator is stationary by design (level/gear/allocations all frozen, exp applied after), so growing *only* the locked base would be internally inconsistent. Documented, with #1253 filed for the redesign question.
- **Live frontend wiring included, not deferred** to #1225. Right — unlike the (empty-until-content) proficiency precedent, every character has a class, so the live battler must compose the locked base or the replay diverges. The minimal `PlayerData.LockedBaseDistribution` delivery is the correct surface for that.

### What's good

- The locked base is derived from `(class, level)` and never stored — it re-tunes for free when growth vectors change and structurally can't be stripped, which delivers the "floor" with no separate validation rule.
- Composition order made an explicit, mirrored, documented parity contract rather than an accident.
- Domain logic stays in `Game.Core` (`BattleSnapshot`), `BattleService` only threads resolvers, the API layer only maps — layering is clean throughout.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by ginderjeremiah** [comment]
Thanks for the thorough review — all findings addressed in `0391318` (and merged `main`):

**Should-fix 1 — decimal-vs-double parity hole:** fixed at the shared source rather than papered over. `AttributeDistribution.GetDistributionModifier` now evaluates `BaseAmount + AmountPerLevel × level` in `double` (casting each `decimal` operand first) instead of decimal-then-cast, so it's bit-identical to the frontend's IEEE-754 recompute for *any* authored value — covering the **enemy** distribution (also client-recomputed) as well as the class locked base. Added a `fractionalPerLevel` parity vector to both suites — `0.1 × 3`, which the old path produced as `0.3` vs the correct double `0.30000000000000004` — so the case that would've diverged is now pinned green. Documented in `docs/backend.md` + the module/test comments (out of the test-comment-only constraint).

**Should-fix 2 — inconsistent failure policy:** `LoginController.BuildPlayerData` now **fails loud** on an unresolvable class, matching `BattleService.ResolveClass` / `AccountService`. That surfaced one fixture (`TagsControllerTests`) that seeded its player's class directly without a cache reload — added the reload (mirroring the battle-test pattern), so the fail-loud path is exercised correctly.

**Test gap:** added the locked-base change-detection case (reassign → re-derive) and an explicit order assertion that `reset` receives `[...lockedBase, ...proficiency]` with **both** non-empty.

**Nits:** widened the `AttributeDistribution` contract doc (now reused for the class fingerprint); branch updated with `main` (proficiency-screen rework).

Left as-is per your sign-off: offline base frozen at window-start (#1253 tracks the redesign question) and the live-wiring-not-deferred call. Full backend (Core 828 / Application 583 / Api 615) + frontend (2556, coverage held) + lint + format green locally; CI re-running.

---
_Generated by [Claude Code](https://claude.ai/code)_
